[LoopDist] Consider reads and writes together for runtime checks

mcberg2021 opened the pull request:
Emit safety guards for ptr accesses when cross partition loads exist which have a corresponding store to the same address in a different partition.  This will emit the necessary ptr checks for these accesses.

The test case was obtained from SuperTest, which SiFive runs regularly.  We enabled LoopDistribution by default in our downstream compiler, this change was part of that enablement.

mcberg2021 committed:
[LoopDist] Add some runtime checks for cross partition loads

Emit safety guards for ptr accesses when cross partition
loads exist which have a corresponding store to the same
address in a different partition.  This will emit the
necessary ptr checks for these accesses.

mcberg2021 commented:
Ping

mcberg2021 commented on llvm/test/Transforms/LoopDistribute/cross-partition-access.ll:
```diff
@@ -0,0 +1,207 @@
+; CHECK-NEXT:    ret void
+;
+entry:
+  %cmp = icmp sgt i32 %len, 0
+  br i1 %cmp, label %if.then, label %if.end
+
+if.then:                                          ; preds = %entry
+  %sub.neg = add nsw i32 %len, -2147483647
+  %cmp1.not51 = icmp eq i32 %len, 1
+  br i1 %cmp1.not51, label %if.end, label %for.body.preheader
+
+for.body.preheader:                               ; preds = %if.then
+  %add = sub nuw i32 -2147483648, %len
+  %i = zext nneg i32 %add to i64
+  br label %for.body
```
I'll look into simplifying it.

mcberg2021 commented on llvm/lib/Transforms/Scalar/LoopDistribute.cpp:
```diff
@@ -521,6 +521,21 @@ class InstPartitionContainer {
+        if (Ptr->hasOneUse() || Partition == -1)
+          continue;
+
+        for (User *U : Ptr->users())
```
I'm going to guess at what you mean.  The original code handles all the like type ptrs from store to store and load to load in all partitions, so I believe that is covered.  I could make this a bit more generic in that I could use RtPtrCheck->Pointers[I].IsWritePtr to flip what we are looking for, adding a store check to based on the opposite state.  That would cover loads with cross partition stores as well as stores with cross partition loads.  If other, please elaborate.

Meinersbur replied:
There are other instructions other than LoadInst that only read from memory, such as calls marked with `ModRefInfo::Ref`. I think this should use `LAI.getInstructionsForAccess(Ptr, false)` as the code above, best even integrated with the code above: If `RtPtrCheck->Pointers[I].IsWritePtr` is true, compare `LAI.getInstructionsForAccess(Ptr, false)` and `LAI.getInstructionsForAccess(Ptr, true)`

mcberg2021 commented on llvm/test/Transforms/LoopDistribute/cross-partition-access.ll:
```diff
@@ -0,0 +1,207 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -aa-pipeline=basic-aa -passes=loop-simplify,loop-distribute -enable-loop-distribute -verify-loop-info -verify-dom-info -S %s | FileCheck %s
```
Sure If I collect the output of loop-simplify and use that for IR/meta data for this test case.

mcberg2021 committed:
Merge branch 'main' into users/mcberg2021_LoopDistribute

Meinersbur commented on llvm/lib/Transforms/Scalar/LoopDistribute.cpp:
```diff
@@ -521,6 +521,21 @@ class InstPartitionContainer {
+          continue;
+
+        for (User *U : Ptr->users())
+          if (auto *CurLoad = dyn_cast<LoadInst>(U))
+            if (L->contains(CurLoad->getParent()))
+              if (Partition != (int)this->InstToPartitionId[CurLoad]) {
```
```suggestion
          auto *CurLoad = dyn_cast<LoadInst>(U);
          if (!CurLoad)
            continue; 
          if (!L->contains(CurLoad->getParent()))
            continue;
          if (Partition == (int)this->InstToPartitionId[CurLoad]) 
            continue;
```
[Use Early Exits and continue to Simplify Code](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code)

mcberg2021 committed:
Merge branch 'main' into users/mcberg2021_LoopDistribute

mcberg2021 committed:
[LoopDist] Consider reads and writes together for runtime checks

Add back in getInstructionsForAccess as query for Instructions modifying
memory to mark cross partition accesses.

Meinersbur reviewed (commented):
Why make it more complicated than necessary?

Meinersbur reviewed (approved):
LGTM, thank you

Meinersbur commented:
Could you make the title more concrete before landing? Such as "Also consider reads during partitioning"
